sql/src/main/java/io/crate/executor/transport/TransportFetchNodeAction.java

```diff
@@ -111,60 +108,60 @@ public String executorName() {
 
     @Override
     public void nodeOperation(final NodeFetchRequest request,
-                               final ActionListener<NodeFetchResponse> fetchResponse) {
-        statsTables.operationStarted(request.executionPhaseId(), request.jobId(), "fetch");
-        String ramAccountingContextId = String.format(Locale.ENGLISH, "%s: %d", request.jobId(), request.executionPhaseId());
+                              final ActionListener<NodeFetchResponse> fetchResponse) {
+
+        String ramAccountingContextId = String.format(Locale.ENGLISH, "%s: %d", request.jobId(), request.fetchPhaseId());
         final RamAccountingContext ramAccountingContext = new RamAccountingContext(ramAccountingContextId, circuitBreaker);
+        try {
+            statsTables.operationStarted(request.fetchPhaseId(), request.jobId(), "fetch");
 
-        NodeFetchOperation fetchOperation = new NodeFetchOperation(
-                request.jobId(),
-                request.executionPhaseId(),
-                request.jobSearchContextDocIds(),
-                request.toFetchReferences(),
-                jobContextService,
-                threadPool,
-                functions,
-                ramAccountingContext);
-
-        Streamer<?>[] streamers = outputStreamers(request.toFetchReferences());
-        SingleBucketBuilder bucketBuilder = new SingleBucketBuilder(streamers);
-        final NodeFetchResponse response = new NodeFetchResponse(streamers);
-        Futures.addCallback(bucketBuilder.result(), new FutureCallback<Bucket>() {
-            @Override
-            public void onSuccess(@Nullable Bucket result) {
-                assert result != null;
-                response.rows(result);
-
-                fetchResponse.onResponse(response);
-                statsTables.operationFinished(request.executionPhaseId(), null, ramAccountingContext.totalBytes());
-                ramAccountingContext.close();
-            }
+            JobExecutionContext jobExecutionContext = jobContextService.getContext(request.jobId());
+            final FetchContext fetchContext = jobExecutionContext.getSubContext(request.fetchPhaseId());
 
-            @Override
-            public void onFailure(@Nonnull Throwable t) {
-                fetchResponse.onFailure(t);
-                statsTables.operationFinished(request.executionPhaseId(), Exceptions.messageOf(t),
-                        ramAccountingContext.totalBytes());
-                ramAccountingContext.close();
+            // nothing to fetch, just close
+            if (request.toFetch() == null) {
```

Comment:
could it be that toFetch is empty instead of null in which case the shortcut here would also make sense?